File: Test/PyBind/scripts/advanced/AsyncPolling.py
from sys import argv
import threading
from threading import Thread
from time import sleep

from D14UIKit import *

DEMO_NAME = 'AsyncPolling'

# These variables need to be promoted to global definition;
# otherwise, the child thread cannot locate them.
curr = 0
done = False

def main():
    dpi = 96.0
    if len(argv) >= 2 and argv[1] == 'HighDPI':
        dpi = 192.0

    app = Application(DEMO_NAME, dpi)
    app.size = Size(600, 266)

    #------------------------------------------- Initialize UI objects.

    mwnd = MainWindow(DEMO_NAME)

    clntArea = Panel()
    mwnd.content = clntArea

    startBtn = FilledButton()
    startBtn.parent = clntArea
    startBtn.size = Size(200, 50)
    startBtn.position = Point(50, 50)
    startBtn.roundRadius = 5
    startBtn.text ='Start Progress'

    controlBtn = FilledButton()
    controlBtn.parent = clntArea
    controlBtn.size = Size(200, 50)
    controlBtn.position = Point(275, 50)
    controlBtn.roundRadius = 5
    controlBtn.text ='Pause Progress'

    full = 500
    color = app.themeColor

    progressLbl = Label()
    progressLbl.parent = clntArea
    progressLbl.size = Size(50, 50)
    progressLbl.position = Point(500, 50)
    progressLbl.text = f'0/{full}'
    progressLbl.font = Font('Default/Normal/14')
    progressLbl.horzAlign = Label.HCenter

    progressArea = Panel()
    progressArea.parent = clntArea
    progressArea.size = Size(full, 30)
    progressArea.position = Point(50, 150)
    progressArea.outlineWidth = 2
    progressArea.outlineColor = color
    progressArea.outlineOpacity = 0.5
    progressArea.roundRadius = 5

    challengeLbl = Label()
    challengeLbl.parent = progressArea
    challengeLbl.position = Point(10, 0)
    challengeLbl.height = 30
    challengeLbl.text = 'Challenge: Use the Pause/Resume button to get a 499/500!'
    challengeLbl.fontSize = 12

    progressBar = Panel()
    progressBar.parent = clntArea
    progressBar.moveAbove(progressArea)
    progressBar.size = Size(0, 30)
    progressBar.position = Point(50, 150)
    progressBar.color = color
    progressBar.opacity = 0.5
    progressBar.roundRadius = 5

    #------------------------------------------- Set UI event callbacks.

    work = threading.Event()
    work.set()

    def workThread():
        global curr, done

        for i in range(1, full+1):
            work.wait()
            sleep(0.002)
            curr = i

        done = True

    def startProgress(clkp, e):
        global curr, done

        if e.left:
            startBtn.enabled = False
            # render continuously for polling
            progressBar.animState = True

            # new round: reset thread state
            curr = 0
            done = False
            progressBar.width = 0

            Thread(target=workThread, daemon=True).start()

    startBtn.f_onMouseButtonRelease = startProgress

    def controlProgress(clkp, e):
        if e.left:
            if work.is_set():
                work.clear()
                controlBtn.text = 'Resume Progress'
            else:
                work.set()
                controlBtn.text = 'Pause Progress'

    controlBtn.f_onMouseButtonRelease = controlProgress

    #############################
    # Polling in each frame !!! #
    #############################

    def updateLbl(p):
        progressLbl.text = f'{curr}/{full}'

    progressLbl.f_onUpdate = updateLbl

    def updateBar(p):
        if done:
            startBtn.enabled = True
            progressBar.animState = False
        else:
            progressBar.width = curr

    progressBar.f_onUpdate = updateBar

    exit(app.run())

if __name__ == '__main__':
    main()

File: Test/sources/advanced/AsyncPolling.cpp
#include <atomic>
#include <chrono>
#include <thread>

using namespace std::chrono_literals;

#include "Application.h"
#include "Callback.h"
#include "FilledButton.h"
#include "Label.h"
#include "MainWindow.h"

using namespace d14uikit;

#define DEMO_NAME L"AsyncPolling"

int main(int argc, char* argv[])
{
    float dpi = 96.0f;
    if (argc >= 2 && strcmp(argv[1], "HighDPI") == 0)
    {
        dpi = 192.0f;
    }
    Application app(DEMO_NAME, dpi);
    app.setSize({ 600, 266 });

    //------------------------------------------- Initialize UI objects.

    MainWindow mwnd(DEMO_NAME);

    Panel clntArea;
    mwnd.setContent(&clntArea);

    FilledButton startBtn;
    startBtn.setParent(&clntArea);
    startBtn.setSize({ 200, 50 });
    startBtn.setPosition({ 50, 50 });
    startBtn.setRoundRadius(5);
    startBtn.setText(L"Start Progress");

    FilledButton controlBtn;
    controlBtn.setParent(&clntArea);
    controlBtn.setSize({ 200, 50 });
    controlBtn.setPosition({ 275, 50 });
    controlBtn.setRoundRadius(5);
    controlBtn.setText(L"Pause Progress");

    constexpr int full = 500;
    auto color = app.themeStyle().color;

    Label progressLbl;
    progressLbl.setParent(&clntArea);
    progressLbl.setSize({ 50, 50 });
    progressLbl.setPosition({ 500, 50 });
    progressLbl.setText(L"0/" + std::to_wstring(full));
    auto font = Font(L"Default/Normal/14");
    progressLbl.setFont(&font);
    progressLbl.setHorzAlign(Label::HCenter);

    Panel progressArea;
    progressArea.setParent(&clntArea);
    progressArea.setSize({ full, 30 });
    progressArea.setPosition({ 50, 150 });
    progressArea.setOutlineWidth(2);
    progressArea.setOutlineColor(color);
    progressArea.setOutlineOpacity(0.5f);
    progressArea.setRoundRadius(5);

    Label challengeLbl;
    challengeLbl.setParent(&progressArea);
    challengeLbl.setPosition({ 10, 0 });
    challengeLbl.setHeight(30);
    challengeLbl.setText(L"Challenge: Use the Pause/Resume button to get a 499/500!");
    challengeLbl.setFontSize(12);

    Panel progressBar;
    progressBar.setParent(&clntArea);
    progressBar.moveAbove(&progressArea);
    progressBar.setSize({ 0, 30 });
    progressBar.setPosition({ 50, 150 });
    progressBar.setColor(color);
    progressBar.setOpacity(0.5f);
    progressBar.setRoundRadius(5);

    //------------------------------------------- Set UI event callbacks.

    std::atomic<int> curr = 0;
    std::atomic<bool> work = true;
    std::atomic<bool> done = false;

    startBtn.D14_onMouseButtonRelease(clkp, e, &)
    {
        if (e->left())
        {
            startBtn.setEnabled(false);
            // render continuously for polling
            progressBar.setAnimState(true);

            // new round: reset thread state
            curr = 0;
            work.notify_one();
            done = false;
            progressBar.setWidth(0);

            std::thread([&]
            {
                for (int i = 1; i <= full; ++i)
                {
                    work.wait(false);
                    std::this_thread::sleep_for(2ms);
                    curr = i;
                }
                done = true;
            })
            .detach();
        }
    };
    controlBtn.D14_onMouseButtonRelease(clkp, e, &)
    {
        if (e->left())
        {
            if (work)
            {
                work = false;
                controlBtn.setText(L"Resume Progress");
            }
            else
            {
                work = true;
                controlBtn.setText(L"Pause Progress");
            }
            work.notify_one();
        }
    };

    ///////////////////////////////
    // Polling in each frame !!! //
    ///////////////////////////////

    progressLbl.D14_onUpdate(p, &)
    {
        progressLbl.setText(std::to_wstring(curr) + L"/" + std::to_wstring(full));
    };
    progressBar.D14_onUpdate(p, &)
    {
        if (done)
        {
            startBtn.setEnabled(true);
            progressBar.setAnimState(false);
        }
        else progressBar.setWidth(curr);
    };

    return app.run();
}
